File: android/src/main/java/com/margelo/nitro/backgroundlocationrelay/ConfigStore.kt
package com.margelo.nitro.backgroundlocationrelay

import android.content.Context
import org.json.JSONObject

object ConfigStore {
  private const val PREFS_NAME = "BackgroundLocationRelay"
  private const val CONFIG_KEY = "config"
  private const val WAS_RUNNING_KEY = "was_running"

  fun save(context: Context, config: BackgroundLocationRelayConfig) {
    context
      .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
      .edit()
      .putString(CONFIG_KEY, toJson(config).toString())
      .apply()
  }

  fun setWasRunning(context: Context, running: Boolean) {
    context
      .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
      .edit()
      .putBoolean(WAS_RUNNING_KEY, running)
      .apply()
  }

  fun wasRunning(context: Context): Boolean {
    return context
      .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
      .getBoolean(WAS_RUNNING_KEY, false)
  }

  fun load(context: Context): BackgroundLocationRelayConfig? {
    val raw =
      context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).getString(CONFIG_KEY, null)
        ?: return null

    return try {
      fromJson(JSONObject(raw))
    } catch (error: Exception) {
      RelayLogger.error("Failed to load persisted config: ${error.message}")
      null
    }
  }

  private fun toJson(config: BackgroundLocationRelayConfig): JSONObject {
    val location = config.location
    val request = config.request

    val locationJson =
      JSONObject()
        .put("interval", location.interval)
        .put("fastestInterval", location.fastestInterval)
        .put("distanceFilter", location.distanceFilter)
        .put("accuracy", location.accuracy?.name?.lowercase())
        .put(
          "pausesLocationUpdatesAutomatically",
          location.pausesLocationUpdatesAutomatically,
        )
        .put(
          "showsBackgroundLocationIndicator",
          location.showsBackgroundLocationIndicator,
        )

    val requestJson =
      JSONObject()
        .put("endpoint", request.endpoint)
        .put("headersJson", request.headersJson)
        .put("timeout", request.timeout)
        .put("bodyTemplate", request.bodyTemplate)

    return JSONObject().put("location", locationJson).put("request", requestJson)
  }

  private fun fromJson(json: JSONObject): BackgroundLocationRelayConfig {
    val locationJson = json.getJSONObject("location")
    val requestJson = json.getJSONObject("request")

    val accuracyValue = locationJson.optString("accuracy", "")
    val accuracy =
      if (accuracyValue.isEmpty()) {
        null
      } else {
        LocationAccuracy.values().firstOrNull {
          it.name.equals(accuracyValue, ignoreCase = true)
        }
      }

    return BackgroundLocationRelayConfig(
      location =
        LocationConfig(
          interval = locationJson.getDouble("interval"),
          fastestInterval = locationJson.optNullableDouble("fastestInterval"),
          distanceFilter = locationJson.optNullableDouble("distanceFilter"),
          accuracy = accuracy,
          pausesLocationUpdatesAutomatically =
            locationJson.optNullableBoolean("pausesLocationUpdatesAutomatically"),
          showsBackgroundLocationIndicator =
            locationJson.optNullableBoolean("showsBackgroundLocationIndicator"),
        ),
      request =
        RequestConfig(
          endpoint = requestJson.getString("endpoint"),
          headersJson = requestJson.optNullableString("headersJson"),
          timeout = requestJson.optNullableDouble("timeout"),
          bodyTemplate = requestJson.getString("bodyTemplate"),
        ),
    )
  }

  private fun JSONObject.optNullableDouble(key: String): Double? {
    if (!has(key) || isNull(key)) {
      return null
    }
    return getDouble(key)
  }

  private fun JSONObject.optNullableBoolean(key: String): Boolean? {
    if (!has(key) || isNull(key)) {
      return null
    }
    return getBoolean(key)
  }

  private fun JSONObject.optNullableString(key: String): String? {
    if (!has(key) || isNull(key)) {
      return null
    }
    val value = getString(key)
    return if (value.isEmpty()) null else value
  }
}

File: ios/ConfigStore.swift
import Foundation

enum ConfigStore {
  private static let configKey = "BackgroundLocationRelay.config"
  private static let wasRunningKey = "BackgroundLocationRelay.wasRunning"

  static func setWasRunning(_ running: Bool) {
    UserDefaults.standard.set(running, forKey: wasRunningKey)
  }

  static func wasRunning() -> Bool {
    UserDefaults.standard.bool(forKey: wasRunningKey)
  }

  static func save(_ config: BackgroundLocationRelayConfig) {
    let payload: [String: Any] = [
      "location": locationDictionary(config.location),
      "request": requestDictionary(config.request),
    ]

    UserDefaults.standard.set(payload, forKey: configKey)
  }

  static func load() -> BackgroundLocationRelayConfig? {
    guard let payload = UserDefaults.standard.dictionary(forKey: configKey) else {
      return nil
    }

    guard
      let locationPayload = payload["location"] as? [String: Any],
      let requestPayload = payload["request"] as? [String: Any]
    else {
      return nil
    }

    return BackgroundLocationRelayConfig(
      location: locationConfig(from: locationPayload),
      request: requestConfig(from: requestPayload)
    )
  }

  private static func locationDictionary(_ location: LocationConfig) -> [String: Any] {
    var payload: [String: Any] = ["interval": location.interval]

    if let fastestInterval = location.fastestInterval {
      payload["fastestInterval"] = fastestInterval
    }
    if let distanceFilter = location.distanceFilter {
      payload["distanceFilter"] = distanceFilter
    }
    if let accuracy = location.accuracy {
      payload["accuracy"] = accuracy.stringValue
    }
    if let pauses = location.pausesLocationUpdatesAutomatically {
      payload["pausesLocationUpdatesAutomatically"] = pauses
    }
    if let showsIndicator = location.showsBackgroundLocationIndicator {
      payload["showsBackgroundLocationIndicator"] = showsIndicator
    }

    return payload
  }

  private static func requestDictionary(_ request: RequestConfig) -> [String: Any] {
    var payload: [String: Any] = [
      "endpoint": request.endpoint,
      "bodyTemplate": request.bodyTemplate,
    ]

    if let headersJson = request.headersJson {
      payload["headersJson"] = headersJson
    }
    if let timeout = request.timeout {
      payload["timeout"] = timeout
    }

    return payload
  }

  private static func locationConfig(from payload: [String: Any]) -> LocationConfig {
    LocationConfig(
      interval: payload["interval"] as? Double ?? 60_000,
      fastestInterval: payload["fastestInterval"] as? Double,
      distanceFilter: payload["distanceFilter"] as? Double,
      accuracy: (payload["accuracy"] as? String).flatMap(LocationAccuracy.init),
      pausesLocationUpdatesAutomatically:
        payload["pausesLocationUpdatesAutomatically"] as? Bool,
      showsBackgroundLocationIndicator:
        payload["showsBackgroundLocationIndicator"] as? Bool
    )
  }

  private static func requestConfig(from payload: [String: Any]) -> RequestConfig {
    RequestConfig(
      endpoint: payload["endpoint"] as? String ?? "",
      headersJson: payload["headersJson"] as? String,
      timeout: payload["timeout"] as? Double,
      bodyTemplate: payload["bodyTemplate"] as? String ?? ""
    )
  }
}
